### ISSUE TITLE
Support using `?` on types returning Option

### ISSUE THREAD

**Issue by thomcc** [google/evcxr#116]
It would be nice if things like this worked:
```
>> [1, 2, 3].get(4)?
                   ^ `std::option::NoneError` cannot be formatted with the default formatter
`std::option::NoneError` doesn't implement `std::fmt::Display`
help: the trait `std::fmt::Display` is not implemented for `std::option::NoneError`
```

**Comment by davidlattimore** [comment]
That will work if you switch formatting of errors to use the Debug trait instead of Display. You can do that with `:efmt {:?}`

**Comment by thomcc** [comment]
Ah, that works, and I can put it in my init.evcxr to get it permanently. It not working by default is a little unfortunate, but I guess that's your call.

**Comment by davidlattimore** [comment]
I agree it's unfortunate. I think while I was developing support for `?` I switched back and forth a couple of times between using Display and Debug. Not completely sure why I settled on Display. It's possible that whatever crate I was experimenting with didn't implement Debug for its errors. Or maybe I just thought the Display output looked better. I'm open to changing, although don't know how to make an informed decision as to which to use. It's likely to be a trade-off either way. You're right though that Option is a pretty important use-case.

The [docs for Error](https://doc.rust-lang.org/std/error/trait.Error.html) say:

> Errors must describe themselves through the Display and Debug traits

Perhaps there's a case to be made that NoneError should implement Display given what the docs say - although it doesn't look like NoneError implements the Error trait, so perhaps that doesn't apply.

I could attempt to detect the error and either advise the user to switch the error formatter, or switch automatically (although that could potentially be surprising)

**Comment by davidlattimore** [comment]
Ah, found a [discussion](https://github.com/rust-lang/rust/issues/46871) of NoneError not implementing Error. Sounds like the future of NoneError is still a bit uncertain.

**Comment by thomcc** [comment]
Yes, my understanding is that `NoneError` is an unfortunate artifact of how the unstable `Try` trait is defined, and is likely to be removed (IIRC it's necessity is even part of why Try is still unstable).

I think Debug tends to be more often implemented, and e.g. `std::result::Result` requires `Debug` for `expect`, `unwrap` and friends to work (see https://doc.rust-lang.org/src/core/result.rs.html#936-963).

That said `Display` tends to have prettier output, I agree -- errors often just `#[derive(Debug)]` and give user friendly messages in Display.

**Comment by scottmcm** [comment]
> IIRC it's necessity is even part of why Try is still unstable

You recall correctly 🙂 

I figured I'd drop a note here about now-accepted https://github.com/rust-lang/rfcs/pull/3058 -- `NoneError` will [disappear](https://github.com/rust-lang/rust/issues/46871#issuecomment-827103631) from this error message.
